marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs

```diff
@@ -0,0 +1,268 @@
+  -- ^ Specifies that the transaction must send the given assets to the given
+  -- address.
+  --
+  -- Rules to check:
+  --   1. The total assets sent to the address cover at least the specified amount.
+  --   2. Let t be the transaction body, c be the constraint set, and p be the
+  --      PayAssets constraint from c under consideration. If t satisfies c, then
+  --      t' satisfies c', where t' is t with the required assets of p removed
+  --      from the total output to the address of p and c' is c with p removed.
+  | SendToMarloweScript Chain.Assets (Core.Datum v)
+  -- ^ Specifies that the transaction must send the given assets and the given
+  --   datum to a Marlowe script address of the correct version.
+  --
+  -- Rules to check:
+  --   1. The transaction sends an output with the given assets and datum to a
+  --      script address.
+  --   2. The script address in rule 1 is in the script address set for the
+  --      corresponding Marlowe version.
+  | SendToPayoutScript Chain.Assets (Core.PayoutDatum v)
+  -- ^ Specifies that the transaction must send the given assets and the given
+  --   datum to the Payout script address associated with the Marlowe script
+  --   address of the contract.
+  --
+  -- Rules to check:
+  --   1. The transaction sends an output with the given assets and datum to a
+  --      script address.
+  --   2. The transaction consumes an output from a marlowe script address.
+  --   2. The script address in rule 1 is the payout script address associated
+  --      with the marlowe script address in rule 2.
+
+-- | Require the transaction to send the specified assets to the address.
+mustPayToAddress :: Core.IsMarloweVersion v => Chain.Assets -> Chain.Address -> TxConstraints v
+mustPayToAddress assets address = mempty { outputConstraints = Set.singleton $ PayAssets assets address }
+
+-- | Require the transaction to send an output to the marlowe script address
+-- with the given assets and the given datum.
+mustSendMarloweOutput :: Core.IsMarloweVersion v => Chain.Assets -> Core.Datum v -> TxConstraints v
+mustSendMarloweOutput assets datum =
+  mempty { outputConstraints = Set.singleton $ SendToMarloweScript assets datum }
+
+-- | Require the transaction to send an output to the payout script address
+-- with the given assets and the given datum.
+mustSendPayoutOutput :: Core.IsMarloweVersion v => Chain.Assets -> Core.PayoutDatum v -> TxConstraints v
+mustSendPayoutOutput assets datum =
+  mempty { outputConstraints = Set.singleton $ SendToPayoutScript assets datum }
+
+data InputConstraint v
+  = NoInput
+  -- ^ Specifies that the transaction consumes no marlowe or payout script
+  -- inputs. Corresponds to the create transaction type.
+  --
+  -- Rules to check:
+  --   1. All inputs of the transaction do not come from any of the script
+  --      addresses (marlowe or payout) associated with marlowe version v.
+  | MarloweInput Chain.TxOutRef P.POSIXTime P.POSIXTime (Core.Redeemer v)
```

Comment:
Should this include the datum? or will that be looked up from history?

Reply:
Do we need the datum to construct the input? In principle, this should be computable from the existing output, so specifying it here may be redundant.